File: go/euler27.go
package main
import "fmt"

func sieve(n int) []int {
    sieved := make([]bool, n)
    res := []int{}
    for i := 2; i < n; i++ {
        if !sieved[i] {
            res = append(res, i)
            for j := 2 * i; j < n; j += i {
                sieved[j] = true
            }
        }
    }
    return res
}

func isPrime(n int) bool {
    if n <= 1 {
        return false
    }
    for i := 2; i * i <= n; i++ {
        if n % i == 0 {
            return false
        }
    }
    return true
}

func main() {
    res := 0
    max := 1000
    maxStreak := 0
    primes := sieve(max)
    primesSet := make(map[int]bool)
    for _, p := range primes{
        primesSet[p] = true
    }
    
    for _, b := range primes {
        for a := -999; a <= 999; a++ {
            for n := 0; ; n++ {
                curr := n * n + a * n + b
                if curr <= 0 {
                    break
                }
                
                _, found := primesSet[curr]
                if found || (curr > max && isPrime(curr)) {
                    primesSet[curr] = true
                } else {
                    break
                }
                if n > maxStreak {
                    maxStreak = n
                    res = a * b
                }
            }
        }
    }
    fmt.Println(res)
}

File: ts/euler27.ts
function sieve(): number[] {
    const primes: number[] = [];
    const sieved: boolean[] = new Array(1000).fill(false);
    for (let i: number = 2; i < sieved.length; i++) {
        if (!sieved[i]) {
            primes.push(i);
            for (let j: number = i * i; j < sieved.length; j += i) {
                sieved[j] = true;
            }
        }
    }
    return primes;
}

function isPrime(n: number): boolean {
    if (n == 2) return true;
    if (n < 2 || n % 2 == 0) return false;
    for (let i: number = 3; i * i <= n; i += 2) {
        if (n % i == 0) return false;
    }
    return true;
}

function consecutivePrimes(a: number, b: number): number {
    for (let n: number = 0;; n++) {
        if (!isPrime(n * n + a * n + b)) return n;
    }
}

function quadraticPrimes(): number {
    // generating primes with n^2 + an + b
    // starting with n = 0: b must be prime
    let res: number = 0, maxPrimes = 0;
    let primes: number[] = sieve();
    for (let b of primes) {
        for (let a: number = -999; a < 1000; a++) {
            let streak: number = consecutivePrimes(a, b);
            if (streak > maxPrimes) {
                maxPrimes = streak;
                res = a * b;
            }
        }
    }
    return res;
}

console.log(quadraticPrimes());
